airbyte-workers/src/main/java/io/airbyte/workers/process/ExitCodeWatcher.java

```diff
@@ -20,52 +21,111 @@
+  /**
+   * This flag is set to false when we either (a) find the pod's exit code, or (b) when the pod is
+   * deleted. This is so that we call exactly one of onExitCode and onWatchFailure, and we make that
+   * call exactly once.
+   * <p>
+   * We rely on this class being side-effect-free, outside of persistExitCode() and persistFailure().
+   * Those two methods use compareAndSet to prevent race conditions. Everywhere else, we can be sloppy
+   * because we won't actually emit any output.
+   */
+  private final AtomicBoolean active = new AtomicBoolean(true);
 
   /**
-   *
    * @param onExitCode callback used to store the exit code
    * @param onWatchFailure callback that's triggered when the watch fails. should be some failed exit
    *        code.
    */
-  public ExitCodeWatcher(final Consumer<Integer> onExitCode, final Consumer<WatcherException> onWatchFailure) {
+  public ExitCodeWatcher(final String podName,
+                         final String podNamespace,
+                         final Consumer<Integer> onExitCode,
+                         final Runnable onWatchFailure) {
+    this.podName = podName;
+    this.podNamespace = podNamespace;
     this.onExitCode = onExitCode;
     this.onWatchFailure = onWatchFailure;
   }
 
   @Override
-  public void eventReceived(Action action, Pod resource) {
-    try {
-      if (!exitCodeRetrieved && KubePodResourceHelper.isTerminal(resource)) {
-        final ContainerStatus mainContainerStatus = resource.getStatus().getContainerStatuses()
-            .stream()
-            .filter(containerStatus -> containerStatus.getName().equals(KubePodProcess.MAIN_CONTAINER_NAME))
-            .collect(MoreCollectors.onlyElement());
+  public void onAdd(final Pod pod) {
+    if (shouldCheckPod(pod)) {
+      final Optional<Integer> exitCode = getExitCode(pod);
+      exitCode.ifPresent(this::persistExitCode);
+    }
+  }
 
-        if (mainContainerStatus.getState() != null && mainContainerStatus.getState().getTerminated() != null) {
-          final int exitCode = mainContainerStatus.getState().getTerminated().getExitCode();
-          log.info("Processing event with exit code " + exitCode + " for pod: " + resource.getMetadata().getName());
-          onExitCode.accept(exitCode);
-          exitCodeRetrieved = true;
-        }
-      }
-    } catch (Exception e) {
-      String podName = "<unknown_name>";
-      if (resource.getMetadata() != null) {
-        podName = resource.getMetadata().getName();
+  @Override
+  public void onUpdate(final Pod oldPod, final Pod newPod) {
+    if (shouldCheckPod(newPod)) {
+      final Optional<Integer> exitCode = getExitCode(newPod);
+      exitCode.ifPresent(this::persistExitCode);
+    }
+  }
+
+  @Override
+  public void onDelete(final Pod pod, final boolean deletedFinalStateUnknown) {
+    if (shouldCheckPod(pod)) {
+      if (!deletedFinalStateUnknown) {
+        final Optional<Integer> exitCode = getExitCode(pod);
+        exitCode.ifPresentOrElse(
```

Comment:
curious, is this formatting via your auto-format or are you intentionally spacing this?